File: MachineStateService/src/main/java/com/smartlaundromat/machine/model/Machine.java
package com.smartlaundromat.machine.model;

import com.smartlaundromat.machine.model.enums.CommProtocol;
import com.smartlaundromat.machine.model.enums.CycleType;
import com.smartlaundromat.machine.model.enums.MachineStatus;
import com.smartlaundromat.machine.model.enums.MachineType;
import jakarta.persistence.*;
import lombok.*;
import java.time.LocalDateTime;

@Entity
@Table(name = "machines")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Machine {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "machine_id", unique = true, nullable = false, length = 30)
    private String machineId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 10)
    private MachineType type;

    @Column(length = 50)
    @Builder.Default
    private String brand = "LG";

    @Column(length = 50)
    @Builder.Default
    private String model = "Commercial Pro";

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    @Builder.Default
    private MachineStatus status = MachineStatus.IDLE;

    /** Remote-control transport used to start/stop this machine (EQLink, Modbus RTU, or MQTT). */
    @Enumerated(EnumType.STRING)
    @Column(name = "comm_protocol", nullable = false, length = 10)
    @Builder.Default
    private CommProtocol commProtocol = CommProtocol.MQTT;

    @Column(name = "is_online", nullable = false)
    @Builder.Default
    private Boolean isOnline = true;

    @Column(name = "last_heartbeat")
    @Builder.Default
    private LocalDateTime lastHeartbeat = LocalDateTime.now();

    // Current cycle info
    @Enumerated(EnumType.STRING)
    @Column(name = "current_cycle_type", length = 20)
    @Builder.Default
    private CycleType currentCycleType = CycleType.NONE;

    @Column(name = "cycle_started_at")
    private LocalDateTime cycleStartedAt;

    @Column(name = "cycle_duration_minutes")
    private Integer cycleDurationMinutes;

    @Column(name = "cycle_ends_at")
    private LocalDateTime cycleEndsAt;

    @Column(name = "cycle_progress")
    @Builder.Default
    private Integer cycleProgress = 0;

    // Telemetry
    @Column(precision = 5)
    private Double temperature;

    @Column(precision = 5)
    private Double humidity;

    @Column(name = "water_level", precision = 5)
    private Double waterLevel;

    @Column(name = "spin_speed")
    private Integer spinSpeed;

    @Column(precision = 5)
    private Double vibration;

    @Column(name = "door_locked")
    @Builder.Default
    private Boolean doorLocked = false;

    @Column(name = "power_consumption")
    private Double powerConsumption;

    // Error tracking
    @Column(name = "error_code", length = 20)
    private String errorCode;

    @Column(name = "error_message", length = 200)
    private String errorMessage;

    // Maintenance
    @Column(name = "total_cycles")
    @Builder.Default
    private Integer totalCycles = 0;

    @Column(name = "cycles_since_service")
    @Builder.Default
    private Integer cyclesSinceService = 0;

    @Column(name = "last_service_date")
    private LocalDateTime lastServiceDate;

    // Location
    @Column(length = 30)
    @Builder.Default
    private String zone = "main";

    @Column
    @Builder.Default
    private Integer position = 1;

    @Column(name = "created_at", nullable = false, updatable = false)
    @Builder.Default
    private LocalDateTime createdAt = LocalDateTime.now();

    @Column(name = "updated_at")
    @Builder.Default
    private LocalDateTime updatedAt = LocalDateTime.now();

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    public boolean isAvailable() {
        return this.status == MachineStatus.IDLE && this.isOnline && !this.doorLocked;
    }

    public String getDisplayName() {
        String typeLabel = this.type == MachineType.WASHER ? "Washer" : "Dryer";
        String number = this.machineId.replaceAll("\\D+", "");
        return typeLabel + " " + number;
    }
}

File: SmartLaundromatControlSystem/src/models/Machine.js
const mongoose = require('mongoose');

/**
 * Machine Schema - Stores telemetry data for washers and dryers
 * Simulates LG commercial laundry equipment
 */
const machineSchema = new mongoose.Schema({
    machineId: {
        type: String,
        required: true,
        unique: true
    },
    type: {
        type: String,
        enum: ['washer', 'dryer'],
        required: true
    },
    brand: {
        type: String,
        default: 'LG'
    },
    model: {
        type: String,
        default: 'Commercial Pro'
    },
    // Machine operational status
    status: {
        type: String,
        enum: ['IDLE', 'RUNNING', 'PAUSED', 'FINISHED', 'ERROR', 'MAINTENANCE', 'OFFLINE'],
        default: 'IDLE'
    },
    // Current cycle information
    currentCycle: {
        type: {
            type: String,
            enum: ['none', 'quick', 'normal', 'heavy', 'delicate', 'sanitize', 'low_heat', 'medium_heat', 'high_heat'],
            default: 'none'
        },
        startedAt: Date,
        duration: Number, // in minutes
        remainingTime: Number, // in minutes
        progress: { type: Number, default: 0 } // 0-100%
    },
    // Telemetry data
    telemetry: {
        temperature: { type: Number, default: 0 }, // Celsius
        humidity: { type: Number, default: 0 }, // Percentage (for dryers)
        waterLevel: { type: Number, default: 0 }, // Percentage (for washers)
        spinSpeed: { type: Number, default: 0 }, // RPM
        vibration: { type: Number, default: 0 }, // Intensity level 0-10
        doorOpen: { type: Boolean, default: false },
        powerConsumption: { type: Number, default: 0 }, // Watts
        waterUsage: { type: Number, default: 0 } // Liters (for washers)
    },
    // Error information
    errorCode: {
        type: String,
        default: null
    },
    errorMessage: {
        type: String,
        default: null
    },
    // Maintenance tracking
    maintenance: {
        lastServiceDate: Date,
        nextServiceDate: Date,
        totalCycles: { type: Number, default: 0 },
        cyclesSinceService: { type: Number, default: 0 }
    },
    // Connection status
    isOnline: {
        type: Boolean,
        default: true
    },
    lastHeartbeat: {
        type: Date,
        default: Date.now
    },
    // Location
    location: {
        zone: { type: String, default: 'main' },
        position: { type: Number, default: 1 }
    }
}, {
    timestamps: true
});

// Index for efficient queries (machineId index is already created by unique: true)
machineSchema.index({ type: 1, status: 1 });

// Virtual to check if machine is available for use
machineSchema.virtual('isAvailable').get(function() {
    return this.status === 'IDLE' && this.isOnline && !this.telemetry.doorOpen;
});

// Virtual to get display name
machineSchema.virtual('displayName').get(function() {
    const typeLabel = this.type === 'washer' ? 'Washer' : 'Dryer';
    const number = this.machineId.split('_')[1];
    return `${typeLabel} ${number}`;
});

module.exports = mongoose.model('Machine', machineSchema);
